### ISSUE TITLE
make build-builder run component builds in parallel

### ISSUE THREAD

**Issue by baumanj** [habitat-sh/builder#108]
Ideally, we'd like to get done as fast as possible as well as get to the failing component as fast as possible. The issues blocking this are:

1. There's only one `pre_build.env` per package and some components share packages, so the commands to remove them can cause spurious errors
2. There may be races with the use of `sccache`; this needs investigation

Assuming those can be resolved; this code basically does what's needed (some details such as logging elided):
```bash
build-builder() {
  declare -A pids
  for component in "$@"; do
    build-builder-component "$component" &
    pids[$component]=$!
  done

  while true; do
    if wait -n "${pids[*]}"; then
      continue # build-builder-component success
    else
      wait_exit=$?
    fi

    if [[ $wait_exit -eq 127 ]]; then # no more jobs to wait for
      return 0
    else
      # might want to stop the other build jobs here
      return $wait_exit
    fi
  done
}
```

**Comment by fnichol** [comment]
I think this is well worth a try and would at least let us know if a) it's going to be 100% success, b) we need to think about a few point of synchronization (like the `pre_build.env` files, etc), or c) ??

Have you given this a try?

**Comment by baumanj** [comment]
I did try this and I think it was basically working, but I was getting build failures when two build jobs had a shared dependency and the second one would fail when it tried to remove the `pre_build.env` file that didn't exist.

What's the easiest way to iterate on `hab-plan-build.sh`?

**Comment by fnichol** [comment]
Oh, maybe we can make the creation/removal of `pre_build.env` a little more permissive (or loose) which could help here. Aside from that race issue I'd imagine we could iterate in the habitat-sh/builder repo's `.studiorc`--unless I'm missing something

**Comment by chefsalim** [comment]
not a priority and not clear how much time/effort is going to be needed to make this happen reliably - closing for now.
